jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/HibernatePersistenceProviderAdaptor.java

```diff
@@ -54,6 +55,7 @@
 
     @Override
     public void injectJtaManager(JtaManager jtaManager) {
+        WildFlyCustomJtaPlatform.setTransactionSynchronizationRegistry(jtaManager.getSynchronizationRegistry());
```

Comment:
The TSR that this will set is ContextTransactionSynchronizationRegistry.getInstance().

https://github.com/wildfly/wildfly/blob/master/jpa/subsystem/src/main/java/org/jboss/as/jpa/transaction/JtaManagerImpl.java#L49

That TSR will have the same synchronization ordering issue that was dealt with elsewhere in the WFLY-11360 fix.  The WFLY-11360 fix did not deal with JtaManagerImpl as it was fixing problems I had introduced in a previous commit. This use of ContextTransactionSynchronizationRegistry.getInstance() predated that; it came in as part of your work on https://issues.jboss.org/browse/WFLY-10426.

I don't know if this point indicates a new issue introduced by this PR, i.e. whether the existing "jndiService().locate(TSR_NAME)" would have looked up the same object as well.

Reply:
Thanks, will update this also.

Reply:
Good point to check this.  

The TSR passed into WildFlyCustomJtaPlatform.setTransactionSynchronizationRegistry(TransactionSynchronizationRegistry tsr) is an instance of org.wildfly.transaction.client.ContextTransactionSynchronizationRegistry and the jndiService() lookup returns org.jboss.as.txn.service.internal.tsr.TransactionSynchronizationRegistryWrapper which delegates to ContextTransactionSynchronizationRegistry.  So, that should work.

Reply:
Sounds like it doesn't work the same. The jndiService() lookup returns org.jboss.as.txn.service.internal.tsr.TransactionSynchronizationRegistryWrapper which AIUI from WFLY-11360 is the correct object if the JCA synchronization ordering is wanted.  Passing org.wildfly.transaction.client.ContextTransactionSynchronizationRegistry to setTransactionSynchronizationRegistry means all uses of WildFlyCustomJtaPlatform will not be seeing that object and instead will be seeing one that does not do the ordering.

Reply:
To clarify: TransactionSynchronizationRegistryWrapper does delegate to ContextTransactionSynchronizationRegistry, but only after doing the JCAOrderedLastSynchronizationList stuff. Bypassing the wrapper means that is bypassed.

Reply:
WildFlyCustomJtaPlatform is only used to register Synchronization classes from the "org.hibernate" package which JCAOrderedLastSynchronizationList doesn't need to know about (currently).  

Following up on this path, I'm going to investigate if the TSR used by the JPA container could ever be the wrong TSR that doesn't do the JCAOrderedLastSynchronizationList stuff.  Mostly, I'm just checking the org.jboss.as.jpa.injectors.PersistenceContextInjectionSource path to see which TSR is used, as its not clear from reading the code.

Reply:
Looks like the PersistenceContextInjectionSource class is correctly using org.jboss.as.txn.service.internal.tsr.TransactionSynchronizationRegistryWrapper, so all paths that reach the JPA container TransactionUtil class (registers the JPA container tx Synchronization) are correct.

Reply:
I created [https://issues.jboss.org/browse/WFLY-11446] for tracking the different TransactionSynchronizationRegistry returned by JNDI lookups versus what EJB3/JCA/JPA containers use internally.  

I'm going to mark this conversation as resolved.

@tomjenkinson fyi ^